[android] Adding builder pattern for LocationComponent activation

langsmith opened the pull request:
This pr adds a builder to activate the `LocationComponent` and resolves https://github.com/mapbox/mapbox-gl-native/issues/13931.

langsmith commented:
Would love 👁👁  from @mapbox/maps-android  

I'm not sure the best way to handle all of the nasty logic inside [of `activateLocationComponent()`](https://github.com/mapbox/mapbox-gl-native/pull/13941/files#diff-a30fc577bb1ed91a4c5e55630cbc8a46R462). Or should I be managing some of this upstream in `LocationComponentActivationOptions` and/or the builder?

I'm definitely checking `locationComponentActivationOptions.styleRes()` incorrectly at the moment. I noticed that if no style resource is passed through to the `LocationComponentActivationOptions` builder, it's `0` as default. 

At least there are promising signs of (some) things working in the test app 👇 

![ezgif com-resize 1](https://user-images.githubusercontent.com/4394910/52980745-83bcb380-3390-11e9-82fb-6bca7672aadd.gif)

tobrun commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java:
```diff
@@ -422,6 +459,55 @@ public void activateLocationComponent(@NonNull Context context, @NonNull Style s
+
+    Log.d(TAG, "locationComponentActivationOptions context packageName = " + locationComponentActivationOptions.context().getPackageName());
+    Log.d(TAG, "locationComponentActivationOptions style URL = " + locationComponentActivationOptions.style().getUrl());
+    Log.d(TAG, "locationComponentActivationOptions styleRes = " + locationComponentActivationOptions.styleRes());
```
do not log out above and if logging it required use Logger.java instead

langsmith replied:
Was logging as I debugged and played with logic in this method. Removed.

tobrun commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java:
```diff
@@ -422,6 +459,55 @@ public void activateLocationComponent(@NonNull Context context, @NonNull Style s
+    Log.d(TAG, "locationComponentActivationOptions style URL = " + locationComponentActivationOptions.style().getUrl());
+    Log.d(TAG, "locationComponentActivationOptions styleRes = " + locationComponentActivationOptions.styleRes());
+
+    if (locationComponentActivationOptions.context() != null
```
if some data is required we should throw an exception during builder.build if missing, this allows for making these fields non-null.

langsmith replied:
Ok, cool. Have adjusted `build()` to:

```java
public LocationComponentActivationOptions build() {
      if (styleRes != null && locationComponentOptions != null) {
        throw new IllegalArgumentException(
          "You've provided both a style resource and a LocationComponentOptions object. You can't use both and " +
            "you must choose one of the two to style the LocationComponent.");
      }
      if (context != null) {
        throw new NullPointerException(
          "Context in LocationComponentActivationOptions is null");
      }
      if (style != null) {
        throw new NullPointerException(
          "Style in LocationComponentActivationOptions is null. Wait for the map to fully load before passing a ");
      }

      return new LocationComponentActivationOptions(context, style, locationEngine, locationEngineRequest,
        locationComponentOptions, styleRes, useDefaultLocationEngine);
}
```

langsmith replied:
Some refactoring needed, but the checking will happen there

tobrun commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java:
```diff
@@ -422,6 +459,55 @@ public void activateLocationComponent(@NonNull Context context, @NonNull Style s
+      setLocationEngine(null);
+    }
+
+    // TODO: How to do some sort of null check on an int for locationComponentActivationOptions.styleRes() ?
```
By using Integer vs int you can get this behavior though in most cases not the cleanest solution

tobrun commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,170 @@
+
+    public Builder() {
+
+    }
```
Would propose to use one constructor: either go with one that showcases required fields or go with a no args constructor

LukasPaczos replied:
I'd go with the one that takes the required parameters and made the default one private.

langsmith replied:
I don't even think the `no args` constructor is needed. I've removed for now. `Context` and `Style` will always be needed for now, so I've made that the only constructor.

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,170 @@
+ */
+public class LocationComponentActivationOptions {
+
+  private Context context;
```
All of the fields can be `final`.

langsmith replied:
Fixed

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,170 @@
+   *
+   * @return the application's current context
+   */
+  @Nullable
```
Because the field should be final and the `Builder` will require `context`, this getter can be annotated as `NonNull`.

langsmith replied:
Fixed

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,170 @@
+   *
+   * @return the map's fully loaded Style object
+   */
+  @Nullable
```
Same as above.

langsmith replied:
Fixed

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java:
```diff
@@ -422,6 +459,55 @@ public void activateLocationComponent(@NonNull Context context, @NonNull Style s
+        R.style.mapbox_LocationComponent));
+
+    } else if (locationComponentActivationOptions.styleRes() <0
+      && locationComponentActivationOptions.locationComponentOptions() == null
```
You can avoid these double checks by throwing an exception during `Builder#build` if both style res and options are provided.

langsmith replied:
Yep. Have adjusted `build()` to:

```java
 public LocationComponentActivationOptions build() {
      if (styleRes != null && locationComponentOptions != null) {
        throw new IllegalArgumentException(
          "You've provided both a style resource and a LocationComponentOptions object. You can't use both and " +
            "you must choose one of the two to style the LocationComponent.");
      }
      if (context != null) {
        throw new NullPointerException(
          "Context in LocationComponentActivationOptions is null.");
      }
      if (style != null) {
        throw new NullPointerException(
          "Style in LocationComponentActivationOptions is null. Wait for the map to fully load before " +
            "passing the Style object.");
      }

      return new LocationComponentActivationOptions(context, style, locationEngine, locationEngineRequest,
        locationComponentOptions, styleRes, useDefaultLocationEngine);
}
```

langsmith replied:
Some refactoring needed, but the checking will happen there

langsmith commented:
Ok @tobrun and @LukasPaczos . Changes have been made based on your feedback. 

`LocationComponentOptions.createFromAttributes()` seems to be returning a null `LocationComponentOptions`  [here](https://github.com/mapbox/mapbox-gl-native/pull/13941/files#diff-a30fc577bb1ed91a4c5e55630cbc8a46R492) and [here](https://github.com/mapbox/mapbox-gl-native/pull/13941/files#diff-a30fc577bb1ed91a4c5e55630cbc8a46R489). I've got 0️⃣ clue why 🤔. Thoughts?  `Context` isn't null in either of those two spots, so that's not the culprit. Do I have to do more than just pass in `R.style.mapbox_LocationComponent` or `locationComponentActivationOptions.styleRes()`?

LukasPaczos commented:
> `LocationComponentOptions.createFromAttributes()` seems to be returning a null ...

The `NPE` you are seeing is thrown in [here](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java#L593) because you haven't initialized the pieces yet and the `locationLayerController` is `null`. A strategy for initialization via the builder should always be to fetch the context and style, then determine styling options and call `initialize` with those in place. Afterward, set the engine request and the engine itself.

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,184 @@
+        throw new NullPointerException(
+          "Context in LocationComponentActivationOptions is null.");
+      }
+      if (style == null) {
```
You should also check if the style is fully loaded.

langsmith replied:
Ok, I now have:

```java
if (style == null) {
        throw new NullPointerException(
"Style in LocationComponentActivationOptions is null. Make sure the Style object isn't null." +
            " Wait for the map to fully load before passing the Style object to LocationComponentActivationOptions.");
}

if (!style.isFullyLoaded()) {
        throw new IllegalArgumentException(
          "Style in LocationComponentActivationOptions isn't fully loaded. Wait for the map to fully load before " +
            "passing the Style object to LocationComponentActivationOptions.");
}
```

langsmith commented:
@LukasPaczos and @tobrun , refactoring based on your latest review is complete. I've also: 

- adjusted all of the other `location` package test app examples to use the `LocationComponentActivationOptions` builder.

- refactored all of the Espresso tests in `LocationComponentTest` , so that they use the `LocationComponentActivationOptions`

langsmith commented:
crashes are Firebase instrumentation around `LocationEngine` logic 🤔

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,189 @@
+    private int styleRes;
+    private boolean useDefaultLocationEngine;
+
+    public Builder(Context context, Style style) {
```
Missing javadoc and nullability annotations.

langsmith replied:
Ok, I adjusted to:

```java

    /**
     * Constructor for the {@link LocationComponentActivationOptions} builder class.
     * While other activation options are optional, the activation process always requires
     * context and a fully-loaded map {@link Style} object, which is why the two are in this
     * constructor.
     */
    public Builder(@NonNull Context context, @NonNull Style style) {
      this.context = context;
      this.style = style;
    }
```

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,189 @@
+      this.style = style;
+    }
+
+    public Builder locationEngine(LocationEngine locationEngine) {
```
Since we have both an option to pass an engine and a `useDefaultLocationEngine` flag in one place now, we need to document more clearly that `useDefaultLocationEngine` is ignored when a non-null engine is set.

langsmith replied:
Ok, I adjusted to:

```java
    /**
     * Deliver your own {@link LocationEngine} to the LocationComponent.
     * 
     * The true/false 
     * {@link LocationComponentActivationOptions#builder(Context, Style)#useDefaultLocationEngine}
     * activation option is ignored when a non-null {@link LocationEngine} is set via
     * this `locationEngine()` method.
     *
     * @param locationEngine a {@link LocationEngine} object
     *
     * @return The {@link Builder} class with a set LocationEngine.
     */
    public Builder locationEngine(LocationEngine locationEngine) {
      this.locationEngine = locationEngine;
      return this;
    }
```

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,189 @@
+    private LocationEngineRequest locationEngineRequest;
+    private LocationComponentOptions locationComponentOptions;
+    private int styleRes;
+    private boolean useDefaultLocationEngine;
```
This flag should default to `true`. This should also be documented.

langsmith replied:
Ok, I've adjusted to:

```java
/**
     *  Set to true as the default in case a true/false value isn't declared via the builder's
     *  {@link LocationComponentActivationOptions#useDefaultLocationEngine()} method.
     *  
     *  Please be aware that this activation boolean is ignored when a non-null
     *  {@link LocationEngine} is set via the builder's `locationEngine()` method.
     */
    private boolean useDefaultLocationEngine = true;
```

langsmith commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java:
```diff
@@ -314,8 +329,11 @@ public void activateLocationComponent(@NonNull Context context, @NonNull Style s
    * @param context the context
    * @param style   the proxy object for current map style. More info at {@link Style}
    * @param options the options
+   * @deprecated As of 7.2.0,
```
Removed

langsmith commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,189 @@
+ * A class which holds the various options for activating the Maps SDK's {@link LocationComponent} to eventually
+ * show the device location on the map.
+ *
+ * since 7.2.0
```
Removed

langsmith commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,189 @@
+  /**
+   * Builder class for constructing a new instance of {@link LocationComponentActivationOptions}.
+   *
+   * since 7.2.0
```
Removed

langsmith commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,189 @@
+  private final int styleRes;
+  private final boolean useDefaultLocationEngine;
+
+  LocationComponentActivationOptions(@NonNull Context context, @NonNull Style style,
```
fixed

langsmith commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,189 @@
+   *
+   * @return the style resource.
+   */
+  @Nullable
```
Removed

langsmith commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,189 @@
+   * location engine initialized
+   * @return whether the default LocationEngine is used
+   */
+  @Nullable
```
Removed

langsmith commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponent.java:
```diff
@@ -422,6 +458,57 @@ public void activateLocationComponent(@NonNull Context context, @NonNull Style s
+
+    // Set the LocationEngine if one was given to LocationComponentActivationOptions
+    if (locationComponentActivationOptions.locationEngine() != null) {
+      setLocationEngine(locationEngine);
```
Fixed via the `activateLocationComponent()` code you've proposed in a previous review comment

langsmith commented:
Ok @LukasPaczos , this is ready for another round of review. I think the `LocationEngine` logic in `activateLocationComponent` is finally correct and takes into account all of the various builder options. 

All instrumentation tests are finally passing. As we discussed offline, some of the instrumentation tests were passing in `null` for the `LocationEngine`, but I didn't use `.locationEngine(null)` when I refactored the tests to use the builder. My last commit was to add `.locationEngine(null)` to the tests that needed it.

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,209 @@
+  /**
+   * The {@link LocationEngine} which the {@link LocationComponent} should use
+   *
+   * @return the engine, or null if you'd like to only force location updates
```
> or null if you'd like to only force location updates

I think this can be skipped because we can still initialize a default engine.

langsmith replied:
Ok, now reads as 

```
* @return the engine
```

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,209 @@
+   * Builder class for constructing a new instance of {@link LocationComponentActivationOptions}.
+   */
+  public static class Builder {
+    private Context context;
```
NIT this can be final

langsmith replied:
fixed

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,209 @@
+   */
+  public static class Builder {
+    private Context context;
+    private Style style;
```
Same as above.

langsmith replied:
fixed

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,209 @@
+     *
+     * @return The {@link Builder} class with a set LocationEngine.
+     */
+    public Builder locationEngine(LocationEngine locationEngine) {
```
Missing nullability annotations for both the argument and the returned value.

langsmith replied:
Ok, now is

```java
 /**
     * Deliver your own {@link LocationEngine} to the LocationComponent.
     * <p>
     * The true/false
     * {@link LocationComponentActivationOptions#builder(Context, Style)#useDefaultLocationEngine}
     * activation option is ignored when a non-null {@link LocationEngine} is set via
     * this `locationEngine()` method.
     *
     * @param locationEngine a {@link LocationEngine} object
     * @return the {@link Builder} object being constructed
     */
    @NonNull
    public Builder locationEngine(@Nullable LocationEngine locationEngine) {
      this.locationEngine = locationEngine;
      return this;
    }
```

LukasPaczos commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/location/LocationComponentTest.kt:
```diff
@@ -121,22 +133,30 @@ class LocationComponentTest : EspressoTest() {
+        // notNullValue() used for LocationEngine because
+        // com.mapbox.android.core.location.LocationEngineProxy
+        // should be returned, rather than `null`
+        assertThat(locationEngine, notNullValue())
```
This test should set both `locationEngine` to `null` and `useDefaultEngine` to `false` and still test for the `null` location engine.

langsmith replied:
Ok, now the test is 

```kotlin
 @Test
  fun locationComponent_doesntInitializeEngineWhenNullProvided() {
    validateTestSetup()
    val componentAction = object : LocationComponentAction.OnPerformLocationComponentAction {
      override fun onLocationComponentAction(component: LocationComponent, mapboxMap: MapboxMap,
                                             style: Style, uiController: UiController, context: Context) {

          locationComponentActivationOptions = LocationComponentActivationOptions
                  .builder(context, style)
                  .locationEngine(null)
                  .useDefaultLocationEngine(false)
                  .locationComponentOptions(
                          LocationComponentOptions.builder(context)
                                  .staleStateTimeout(200)
                                  .enableStaleState(false)
                                  .accuracyAlpha(.5f)
                                  .accuracyColor(Color.BLUE)
                                  .build()
                  )
                  .build()
        component.activateLocationComponent(locationComponentActivationOptions)

        component.isLocationComponentEnabled = true

        val locationEngine = component.locationEngine
        val componentOptions = component.locationComponentOptions

        assertThat(locationEngine, nullValue())
        assertThat(componentOptions, notNullValue())

        TestingAsyncUtils.waitForLayer(uiController, idlingResource.mapView)
        assertThat(componentOptions?.accuracyAlpha(), `is`(.5f))
        assertThat(componentOptions?.accuracyColor(), `is`(Color.BLUE))
      }
    }

    executeComponentTest(componentAction)
  }
```

langsmith commented:
Ok @LukasPaczos , all feedback from your last review has been addressed.

> It'd be great to add tests verifying if there's an exception thrown when the provided style is not fully loaded and if the useDefaultLocationEngine flag is set to true by default.

https://github.com/mapbox/mapbox-gl-native/pull/13941/commits/21bf960ba959be3786f679af89d87cc9d8f52464#diff-14c9705b4a36f28b8485da302a5703c1


>Adding only .locationEngine(null) is still initializing the default engine, you'll need to add .useDefaultLocationEngine(false) instead, otherwise, the tests might become flaky.

https://github.com/mapbox/mapbox-gl-native/pull/13941/commits/21bf960ba959be3786f679af89d87cc9d8f52464#diff-722a0b240a9c5f84fadee9bb40f558d5

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptions.java:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    /**
+     * @param useDefaultLocationEngine true if you want to initialize and use the
```
Could you add a short note that this is ignored when the non-null engine is set?

langsmith replied:
Ok, now reads as 

```
 /**
     * @param useDefaultLocationEngine true if you want to initialize and use the
     *                                 built-in location engine or false if there
     *                                 should be no location engine initialized
     *                                 This is ignored when null is set as the parameter
     *                                 for {@link LocationComponentActivationOptions#builder
     *                                 (Context, Style)#locationEngine()}.
     * @return the {@link Builder} object being constructed
     */
```

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/test/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptionsTest.java:
```diff
@@ -0,0 +1,135 @@
+    LocationComponentOptions locationComponentOptions = LocationComponentOptions.builder(context)
+      .accuracyAlpha(0.5f)
+      .build();
+    assertNotNull(locationComponentOptions);
```
This has already been verified by another test (`sanity`) and can be removed.

langsmith replied:
Removed

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/test/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptionsTest.java:
```diff
@@ -0,0 +1,135 @@
+    LocationComponentOptions locationComponentOptions = LocationComponentOptions.builder(context)
+      .accuracyAlpha(0.5f)
+      .build();
+    assertNotNull(locationComponentOptions);
```
Same as above.

langsmith replied:
Removed

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/test/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptionsTest.java:
```diff
@@ -0,0 +1,135 @@
+    LocationComponentOptions locationComponentOptions = LocationComponentOptions.builder(context)
+      .accuracyAlpha(0.5f)
+      .build();
+    assertNotNull(locationComponentOptions);
```
Same as above.

langsmith replied:
Removed

LukasPaczos commented on platform/android/MapboxGLAndroidSDK/src/test/java/com/mapbox/mapboxsdk/location/LocationComponentActivationOptionsTest.java:
```diff
@@ -0,0 +1,135 @@
+    LocationComponentOptions locationComponentOptions = LocationComponentOptions.builder(context)
+        .accuracyAlpha(0.5f)
+        .build();
+    assertNotNull(locationComponentOptions);
```
Same as above.

langsmith replied:
Removed

LukasPaczos commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/location/LocationComponentTest.kt:
```diff
@@ -198,13 +224,19 @@ class LocationComponentTest : EspressoTest() {
+
+        locationComponentActivationOptions = LocationComponentActivationOptions
+                  .builder(context, style)
+                .locationEngine(null)
```
This should be `.useDefaultLocationEngine(false)` instead.

langsmith replied:
Ok, replaced `.locationEngine(null)` with `.useDefaultLocationEngine(false)`

LukasPaczos commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/location/LocationComponentTest.kt:
```diff
@@ -228,16 +260,21 @@ class LocationComponentTest : EspressoTest() {
+
+          locationComponentActivationOptions = LocationComponentActivationOptions
+                  .builder(context, style)
+                  .locationEngine(null)
```
Same as above.

langsmith replied:
Resolved

LukasPaczos commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/location/LocationComponentTest.kt:
```diff
@@ -271,13 +308,18 @@ class LocationComponentTest : EspressoTest() {
+
+        locationComponentActivationOptions = LocationComponentActivationOptions
+                .builder(context, style)
+                .locationEngine(null)
```
Same as above.

langsmith replied:
Resolved

LukasPaczos commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/location/LocationComponentTest.kt:
```diff
@@ -303,13 +345,17 @@ class LocationComponentTest : EspressoTest() {
+
+        locationComponentActivationOptions = LocationComponentActivationOptions
+                .builder(context, style)
+                .locationEngine(null)
```
Same as above.

langsmith replied:
Fixed

LukasPaczos commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/location/LocationComponentTest.kt:
```diff
@@ -335,12 +381,16 @@ class LocationComponentTest : EspressoTest() {
+
+        locationComponentActivationOptions = LocationComponentActivationOptions
+                .builder(context, style)
+                .locationEngine(null)
```
Same as above.

langsmith replied:
Fixed

LukasPaczos commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/location/LocationComponentTest.kt:
```diff
@@ -366,12 +416,17 @@ class LocationComponentTest : EspressoTest() {
+
+        locationComponentActivationOptions = LocationComponentActivationOptions
+                .builder(context, style)
+                .locationEngine(null)
```
Same as above.

langsmith replied:
Fixed

LukasPaczos commented on platform/android/MapboxGLAndroidSDKTestApp/src/androidTest/java/com/mapbox/mapboxsdk/location/LocationComponentTest.kt:
```diff
@@ -425,12 +485,17 @@ class LocationComponentTest : EspressoTest() {
+
+        locationComponentActivationOptions = LocationComponentActivationOptions
+                .builder(context, style)
+                .locationEngine(null)
```
Same as above.

langsmith replied:
Fixed

carstenhag commented:
Hey, just FYI: The deprecated notices (what to use instead) are not visible when just opening the file via Android Studio. So I'm only seeing `@Deprecated`, but not what to use instead, like this:

```java
    /** @deprecated */
    @Deprecated
    @RequiresPermission(
        anyOf = {"android.permission.ACCESS_FINE_LOCATION", "android.permission.ACCESS_COARSE_LOCATION"}
    )
    public void activateLocationComponent(@NonNull Context context, @NonNull Style style, @StyleRes int styleRes) {
        this.activateLocationComponent(context, style, LocationComponentOptions.createFromAttributes(context, styleRes));
    }
```

Not sure how this is solved. I would probably need to download the sources, right? There must be a better way! :)

tobrun commented:
@carstenhag the javadoc with `@deprecated` does contain the alternative method to use. There is not much else we can do than follow [standard deprecation setup](https://docs.oracle.com/javase/7/docs/technotes/guides/javadoc/deprecation/deprecation.html). Please let us know if you find another way!
